src/toil/jobStores/fileJobStore.py

```diff
@@ -214,7 +214,15 @@ def updateFile(self, jobStoreFileID, localFilePath):
 
     def readFile(self, jobStoreFileID, localFilePath):
         self._checkJobStoreFileID(jobStoreFileID)
-        shutil.copyfile(self._getAbsPath(jobStoreFileID), localFilePath)
+        if os.stat(self._getAbsPath(jobStoreFileID)).st_dev == \
+                os.stat(os.path.dirname(localFilePath)).st_dev:
+            # The destination could exist hence we should delete before linking. Mirroring behaviour
+            # of shutil.copyfile
+            if os.path.exists(localFilePath):
```

Comment:
Never mind, that wouldn't eliminate the other race removal and recreation via os.link(). But if os.link fails with a succint errno like EEXISTS, can't we eliminate all races with one try: except:?